File: chapter 2 - linked list/2.4.js
const LinkedList = require('./linkedlist');
const Node = require('./node');
const list = new LinkedList();
list.add(1);
list.add(2);
list.add(10);
list.add(5);
list.add(8);
list.add(5);
list.add(3);
const partition = 3;
const lessVal = new LinkedList();
const moreVal = new LinkedList();
let ll = list.head;

while (ll !== null) {
  if (ll.data < partition && ll.data !== partition) {
    lessVal.add(ll.data);
  } else {
    moreVal.add(ll.data);
  }
  ll = ll.next;
}
ll = lessVal.head;
while (ll.next !== null) {
  ll = ll.next;
}

lessVal.listNodes();
console.log('\r\n\r\n');
moreVal.listNodes();

// TODO: To Merge into single LL.

File: chapter 2 - linked list/2.4.py
from linkedlist import LinkedList
from node import Node
def print_nodes(n):
    while(n != None):
        print(n.data)
        n = n.next


def two_buffer_partition(n, partition):
    high_node = None
    low_node = None
    while(n != None):
        new_node = Node(n.data)
        if(n.data < partition):
            if low_node == None:
                low_node = new_node
            else:
                new_node.next = low_node
                low_node = new_node
        else:
            if high_node == None:
                high_node = new_node
            else:
                new_node.next = high_node
                high_node = new_node
        n = n.next
    
    # merging nodes
    final_node = low_node

    # travel to last node
    while(low_node.next != None):
        low_node = low_node.next
    # connect last node of low val to high val node        
    low_node.next = high_node

    # print the final nodes
    print_nodes(final_node)

def partition(n, partition):
    head = n
    tail = n
    while(n != None):
        next = n.next
        print_nodes(next)
        if n.data < partition:
            n.next = head
            head = n
        else:
            tail.next = n
            tail = n
        n=next
    tail.next = None        
    return head

if __name__ == "__main__":
    linked_list = LinkedList()
    linked_list.add(1)
    linked_list.add(2)
    linked_list.add(10)
    linked_list.add(5)
    linked_list.add(8)
    linked_list.add(5)
    linked_list.add(3)
    linked_list.list_nodes()
    print('-------')
    linked_list.head = partition(linked_list.head, 5)
    print('-------')
    two_buffer_partition(linked_list.head, 5)
    print('-------')
    linked_list.list_nodes()
    print('-------')
